### ISSUE TITLE
Screen sharing on Linux with Wayland pops up a select application/screen window twice

### ISSUE THREAD

**Issue by cedricroijakkers** [jitsi/jitsi-meet-electron#963]
## Description
Most Linux distributions are now switching to Wayland from X11, which has a completely different way of handling window security. While in X11, it was perfectly possible for one application to use a part of the display which was being used by another application, in Wayland this now requires user input to approve it first. In Jitsi Meet Electron, this now causes the behaviour that you have to select the application/window twice before the actual screen sharing starts. You first select (and implicitly approve) the screen that you wish to share in Wayland, but then have to select it again in Jisti. By the time you reach the Jitsi selection, there is only one option left to select, so this is useless. As an extra bonus, you only have a few seconds to select the one screen in Jitsi that you want to share, if you are too slow, the Wayland screen selection window pops up again because you did not finish the sharing action. This is not only useless, but also confusing and will cause problems for people with reduced mouse mobility.

## Current behavior
The following screenshots are from GNOME 46 running under Wayland, but they are fairly identical under other desktop environments.

When you click the share button in Jitsi, the following window pops up. This is from [xdg-desktop-portal-gnome](https://gitlab.gnome.org/GNOME/xdg-desktop-portal-gnome) which is the GNOME Wayland integration that implements this functionality, but different desktop environments have a similar implementation:

![image](https://github.com/jitsi/jitsi-meet-electron/assets/15158042/82a88f7d-0c45-4001-8bd0-b99cf9444e67)

You already see the underlying "Share your screen" window coming from Jitsi, but so far this is normal. As you can see I have two screen, so if I select the AOC screen, the popup dissapears and I get dropped into the Jitsi share screen selection:

![image](https://github.com/jitsi/jitsi-meet-electron/assets/15158042/664c2c41-c874-4acc-bc84-1445f4b62b54)

Now I have about 5 seconds to click on the "Application window" tab (fun fact: not the "Your entire screen" tab, even though I selected an entire screen!) and select the one and only application listed in there, and then the screen sharing will start. If I am not quick enough, the desktop portal selection screen pops up again, and I have to repeat everything.

Eventually, if I'm fast enough, it works, but this is not very practical.

## Expected Behavior
Jitsi should detect the usage of Wayland, and should not present the second screen sharing window. Only the desktop portal should be presented, where the user can select a window or a screen. After selecting, the screen share should start immediately in Jitsi.

## Possible Solution
[André Oliveira](https://github.com/SpecialAro) from the [Ferdium](https://github.com/ferdium/ferdium-app) development team recently fixed this issue in Ferdium. If I use the exact same Jitsi server in Ferdium instead of Jitsi Meet Desktop, it works flawlessly. I have also tested Ferdium on X11 and that still works too. Other OSes also work, because Ferdium has built-in detection for Linux and Wayland. Maybe the [pull request where he fixed it](https://github.com/ferdium/ferdium-app/pull/1739) can be used as an inspiration to fix the issue (since Ferdium is also an Electron 30 application).

## Steps to reproduce
See current behaviour.

# Environment details
Any Linux distro with the Wayland window manager.

**Comment by csett86** [comment]
Yes, the screensharing with wayland is suboptimal at the moment. Fullscreen sharing does not work, and the double dialogs exist... see #829 for all the details and history

**Comment by cedricroijakkers** [comment]
I recently saw the upgrade to Electron 30 in Jitsi too, so I thought the hint from Ferdium might be a good idea. André absolutely hit it out of the park in the Ferdium implementation, all kudos to him :smile:

**Comment by saghul** [comment]
I think the trick there is good. If Wayland is used just assume the one and only selected source is the one and don't show the dialog.

**Comment by SpecialAro** [comment]
@csett86, André Oliveira here 😄

Let me know if you need any help in "fixing" (though it is more of an hack than a fix) this!

**Comment by csett86** [comment]
Many thanks here for the Wayland hint @cedricroijakkers and @SpecialAro, @saghul now used this in the jitsi electron-sdk and the new release of the Jitsi Electron App supports it (on alpha.jitsi.net at this moment, but later automatically on meet.jit.si and the next jitsi stable release.

**Comment by saghul** [comment]
There were so many entangled PRs that I forgot to drop a note here! Thank you folks! ❤️

**Comment by cedricroijakkers** [comment]
I have given it a try on alpha.jitsi.net, in the browser it works fine, but in the Jitsi desktop app nothing happens when clicking on the share button. Is this a desktop app issue that needs to be released first or is something else wrong?

**Comment by saghul** [comment]
Are you on 2025.2.0?

That's the one that contains the fix. alpha contains the necessary parts on the other side.

I'll check to make sure it was deployed.

**Comment by cedricroijakkers** [comment]
Aah yes, that was it, the [AUR package](https://aur.archlinux.org/packages/jitsi-meet-desktop-bin) is still on 2025.1.1 so I have built a new one, and indeed now it works fine.

**Comment by csett86** [comment]
Its now part of the latest jitsi meet stable release 2.0.10073 of today: https://github.com/jitsi/jitsi-meet-release-notes/blob/master/CHANGELOG-WEB.md#2010073-2025-02-27
